Sources/StreamVideo/Utils/AudioRecorder/StreamAudioRecorder.swift

```diff
@@ -0,0 +1,180 @@
+        self.audioSession = audioSession
+
+        setUp()
+    }
+
+    deinit {
+        setUpTask?.cancel()
+        setUpTask = nil
+        do {
+            try FileManager.default.removeItem(at: audioRecorderBuilder.fileURL)
+            log.debug("Successfully deleted audio filename")
+        } catch {
+            log.error("Error deleting fileURL.", error: error)
+        }
+    }
+
+    // MARK: - Public API
+
+    /// Starts recording audio asynchronously.
+    open func startRecording() async {
+        await setUpAudioCaptureIfRequired { [weak self] audioRecorder in
+            guard let self, self.hasActiveCall, !audioRecorder.isRecording else {
+                return
+            }
+
+            audioRecorder.record()
+            audioRecorder.isMeteringEnabled = true
+
+            log.debug("️Recording started.")
+            self.updateMetersTimerCancellable = Foundation.Timer
+                .publish(every: 0.1, on: .main, in: .default)
+                .autoconnect()
+                .sink { [weak self, audioRecorder] _ in
+                    Task { [weak self, audioRecorder] in
+                        guard let self else { return }
+                        audioRecorder.updateMeters()
+                        self._metersPublisher.send(audioRecorder.averagePower(forChannel: 0))
+                        log.debug("️Recording meters updated")
+                    }
+                }
+        }
+    }
+
+    /// Stops recording audio asynchronously.
+    open func stopRecording() async {
+        guard let audioRecorder = await audioRecorderBuilder.result, audioRecorder.isRecording else {
+            return
+        }
+
+        updateMetersTimerCancellable?.cancel()
+        updateMetersTimerCancellable = nil
+        audioRecorder.stop()
+        do {
+            try audioSession.setCategory(.playback)
+        } catch {
+            log.error("Failed to set AudiSession category to playback.", error: error)
+        }
+        log.debug("️Recording stopped.")
+    }
+
+    // MARK: - Private helpers
+
+    private func setUp() {
+        setUpTask = Task {
+            do {
+                #if DEBUG
+                let startedOn = Date()
+                try await audioRecorderBuilder.build()
+                let diff = Date().timeIntervalSince1970 - startedOn.timeIntervalSince1970
+                try Task.checkCancellation() // This required to ensure that tests aren't crashing.
+                log.debug("🎙️AVAudioRecorder creation took: \(diff) seconds.")
+                #else
+                try await audioRecorderBuilder.build()
+                #endif
+            } catch {
+                if type(of: error) != CancellationError.self {
+                    log.error("Failed to create AVAudioRecorder.", error: error)
+                }
+            }
+        }
+    }
+
+    private func setUpAudioCaptureIfRequired(
+        _ completionHandler: @escaping (AVAudioRecorder) async -> Void
```

Comment:
why do we need completion handler here? Can't we return the recorder directly after the async stuff finishes?